lib/query-interface.js

```diff
@@ -495,18 +495,20 @@ QueryInterface.prototype.insert = function(instance, tableName, values, options)
   });
 };
 
-QueryInterface.prototype.upsert = function(tableName, values, updateValues, model, options) {
+QueryInterface.prototype.upsert = function(tableName, valuesByField, updateValues, model, options) {
```

Comment:
Correct. The array of values keyed by attribute name is transformed to an array of values keyed by field name before being passed to `QueryInterface`. This is why I renamed the `values` variable to `valuesByField`. The original code then took this, iterated the model attributes and tried to use the attribute name to get data from the `valuesByField` object. This resulted in a undefined which is the whole problem.